lib/rules/no-constant-condition.js

```diff
@@ -115,26 +118,85 @@ module.exports = {
         }
 
         /**
+         * Tracks when the given node contains a constant condition.
+         * @param {ASTNode} node The AST node to check.
+         * @returns {void}
+         * @private
+         */
+        function trackConstantConditionLoop(node) {
+            if (node.test && isConstant(node.test, true)) {
+                loopsInCurrentScope.add(node);
+            }
+        }
+
+        /**
+         * Reports when the set contains the given constant condition node
+         * @param {ASTNode} node The AST node to check.
+         * @returns {void}
+         * @private
+         */
+        function checkConstantConditionLoopInSet(node) {
+            if (loopsInCurrentScope.has(node)) {
+                loopsInCurrentScope.delete(node);
+                context.report({ node, message: "Unexpected constant condition." });
+            }
+        }
+
+        /**
          * Reports when the given node contains a constant condition.
          * @param {ASTNode} node The AST node to check.
          * @returns {void}
          * @private
          */
-        function checkConstantCondition(node) {
+        function reportIfConstant(node) {
             if (node.test && isConstant(node.test, true)) {
                 context.report({ node, message: "Unexpected constant condition." });
             }
         }
 
         /**
+         * Stores current set of constant loops in loopSetStack temporarily
+         * and uses a new set to track constant loops
+         * @returns {void}
+         * @private
+         */
+        function enterFunction() {
+            loopSetStack.push(loopsInCurrentScope);
+            loopsInCurrentScope = new Set();
+        }
+
+        /**
+         * Reports when the set still contains stored constant conditions
+         * @param {ASTNode} node The AST node to check.
+         * @returns {void}
+         * @private
+         */
+        function exitFunction() {
+            loopsInCurrentScope = loopSetStack.pop();
+        }
+
+        /**
          * Checks node when checkLoops option is enabled
          * @param {ASTNode} node The AST node to check.
          * @returns {void}
          * @private
          */
         function checkLoop(node) {
             if (checkLoops) {
-                checkConstantCondition(node);
+                trackConstantConditionLoop(node);
+            }
+        }
+
+        /**
+        * Checks node for yield in init of forloop
+        * @param {ASTNode} node The AST node to check.
+        * @returns {void}
+        * @private
+        */
+        function checkYieldInInitForLoop(node) {
+            if ((node.type === "VariableDeclaration" && node.declarations[0].init.type === "YieldExpression") ||
+                (node.type === "AssignmentExpression" && node.right.type === "YieldExpression")) {
+                context.report({ node: node.parent, message: "Unexpected constant condition." });
```

Comment:
I don't think this is quite correct, because it will miss more complicated uses of `yield`:

```js
for (let foo = 1 + 2 + 3 + yield; bar; baz) {}
```

Have you tried [this approach](https://github.com/eslint/eslint/pull/8762#issuecomment-313025136)?